radix-engine-interface/src/time/date_time.rs

```diff
@@ -0,0 +1,512 @@
+// a) It's a multiple of 400, to make it easier to work with leap years
+// b) We're also shifting the month to 1st March, so that
+//    the extra day on leap years is added to the last month (Feb),
+//    not in the middle of a year (makes some calculations easier)
+const SHIFT_FROM_UNIX_TIME_TO_MARCH_Y2K: i64 = 946684800 + 86400 * (31 + 29);
+const MIN_SUPPORTED_TIMESTAMP: i64 = -62135596800;
+const MAX_SUPPORTED_TIMESTAMP: i64 = 135536014634284799;
+
+#[derive(Encode, Decode, TypeId, PartialEq, Eq, Copy, Clone, Debug)]
+pub struct DateTime {
+    year: u32,
+    month: u8,
+    day_of_month: u8,
+    hour: u8,
+    minute: u8,
+    second: u8,
+}
+
+impl DateTime {
+    pub fn new(
+        year: u32,
+        month: u8,
+        day_of_month: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
+    ) -> Result<Self, String> {
+        if year <= 0 {
+            return Err("Invalid year. Expected a value strictly greater than 0".to_string());
+        }
+
+        if month < 1 || month > 12 {
+            return Err(
+                "Invalid month. Expected a value between 1 (inclusive) and 12 (inclusive)"
+                    .to_string(),
+            );
+        }
+
+        if day_of_month < 1 ||
+            // Check leap year Feb + all other months
+            day_of_month > LEAP_YEAR_DAYS_IN_MONTHS[(month - 1) as usize] ||
+            // Check Feb on non-leap years
+            (!Self::is_leap_year(year) && month == 2 && day_of_month > 28)
+        {
+            return Err("Invalid day of month.\
+                Expected a value between 1 (inclusive)\
+                and, depending on a month, 28, 29 (Feb on a leap year), 30 or 31 (inclusive)"
+                .to_string());
+        }
+
+        if hour > 23 {
+            return Err(
+                "Invalid hour. Expected a value between 0 (inclusive) and 23 (inclusive)"
+                    .to_string(),
+            );
+        }
+
+        if minute > 59 {
+            return Err(
+                "Invalid minute. Expected a value between 0 (inclusive) and 59 (inclusive)"
+                    .to_string(),
+            );
+        }
+
+        if second > 59 {
+            return Err(
+                "Invalid second. Expected a value between 0 (inclusive) and 59 (inclusive)"
+                    .to_string(),
+            );
+        }
+
+        Ok(Self {
+            year,
+            month,
+            day_of_month,
+            hour,
+            minute,
+            second,
+        })
+    }
+
+    pub fn from_instant(instant: &Instant) -> Result<Self, String> {
+        if instant.seconds_since_unix_epoch < MIN_SUPPORTED_TIMESTAMP
+            || instant.seconds_since_unix_epoch > MAX_SUPPORTED_TIMESTAMP
+        {
+            return Err(format!(
+                "Instant out of supported range [{}, {}]",
+                MIN_SUPPORTED_TIMESTAMP, MAX_SUPPORTED_TIMESTAMP
+            ));
+        }
+
```

Comment:
All maths below are guaranteed not to overflow because of the range check above right?

Reply:
correct